File: fe/src/services/company-service.ts
import api from "@/lib/api";
import {
  ApiResponse,
  Company,
  CreateCompanyRequest,
  PaginatedCompanyResponse,
  UpdateCompanyRequest,
} from "@/lib/types";

export async function getCompanies(params?: {
  page?: number;
  limit?: number;
  search?: string;
  sort_by?: string;
  sort_order?: string;
}): Promise<ApiResponse<PaginatedCompanyResponse>> {
  const searchParams = new URLSearchParams();
  if (params?.page) searchParams.set("page", params.page.toString());
  if (params?.limit) searchParams.set("limit", params.limit.toString());
  if (params?.search) searchParams.set("search", params.search);
  if (params?.sort_by) searchParams.set("sort_by", params.sort_by);
  if (params?.sort_order) searchParams.set("sort_order", params.sort_order);
  const query = searchParams.toString() ? `?${searchParams.toString()}` : "";
  const response = await api.get(`/companies${query}`);
  return response.data;
}

export async function getCompanyById(
  id: string
): Promise<ApiResponse<Company>> {
  const response = await api.get(`/companies/${id}`);
  return response.data;
}

export async function createCompany(
  data: CreateCompanyRequest
): Promise<ApiResponse<Company>> {
  const response = await api.post("/companies", data);
  return response.data;
}

export async function updateCompany(
  id: string,
  data: UpdateCompanyRequest
): Promise<ApiResponse<Company>> {
  const response = await api.put(`/companies/${id}`, data);
  return response.data;
}

export async function deleteCompany(id: string): Promise<ApiResponse<null>> {
  const response = await api.delete(`/companies/${id}`);
  return response.data;
}

export async function deleteMultipleCompanies(
  ids: string[]
): Promise<ApiResponse<null>> {
  const response = await api.delete("/companies", { data: { ids } });
  return response.data;
}

File: be/internal/service/company_service.go
package service

import (
	"errors"

	"hris-backend/internal/dto"
	"hris-backend/internal/model"
	"hris-backend/internal/repository"
)

type CompanyService interface {
	GetAll() ([]dto.CompanyResponse, error)
	GetAllPaginated(page, limit int, search, sortBy, sortOrder string) (*dto.PaginatedCompanyResponse, error)
	GetByID(id string) (*dto.CompanyResponse, error)
	Create(req dto.CreateCompanyRequest) (*dto.CompanyResponse, error)
	Update(id string, req dto.UpdateCompanyRequest) (*dto.CompanyResponse, error)
	Delete(id string) error
	DeleteMultiple(ids []string) error
}

type companyService struct {
	companyRepo repository.CompanyRepository
}

func NewCompanyService(companyRepo repository.CompanyRepository) CompanyService {
	return &companyService{companyRepo: companyRepo}
}

func (s *companyService) GetAll() ([]dto.CompanyResponse, error) {
	companies, err := s.companyRepo.FindAll()
	if err != nil {
		return nil, err
	}
	return dto.ToCompanyResponses(companies), nil
}

func (s *companyService) GetAllPaginated(page, limit int, search, sortBy, sortOrder string) (*dto.PaginatedCompanyResponse, error) {
	companies, total, err := s.companyRepo.FindAllPaginated(page, limit, search, sortBy, sortOrder)
	if err != nil {
		return nil, err
	}

	totalPages := int(total) / limit
	if int(total)%limit > 0 {
		totalPages++
	}

	return &dto.PaginatedCompanyResponse{
		Data:       dto.ToCompanyResponses(companies),
		Page:       page,
		Limit:      limit,
		TotalItems: total,
		TotalPages: totalPages,
	}, nil
}

func (s *companyService) GetByID(id string) (*dto.CompanyResponse, error) {
	company, err := s.companyRepo.FindByID(id)
	if err != nil {
		return nil, errors.New("company not found")
	}
	response := dto.ToCompanyResponse(company)
	return &response, nil
}

func (s *companyService) Create(req dto.CreateCompanyRequest) (*dto.CompanyResponse, error) {
	company := &model.Company{
		Name:     req.Name,
		Address:  req.Address,
		Phone:    req.Phone,
		Email:    req.Email,
		NPWP:     req.NPWP,
		NPP:      req.NPP,
		Logo:     req.Logo,
		IsActive: true,
	}

	if err := s.companyRepo.Create(company); err != nil {
		return nil, errors.New("failed to create company")
	}

	response := dto.ToCompanyResponse(company)
	return &response, nil
}

func (s *companyService) Update(id string, req dto.UpdateCompanyRequest) (*dto.CompanyResponse, error) {
	company, err := s.companyRepo.FindByID(id)
	if err != nil {
		return nil, errors.New("company not found")
	}

	if req.Name != "" {
		company.Name = req.Name
	}
	if req.Address != "" {
		company.Address = req.Address
	}
	if req.Phone != "" {
		company.Phone = req.Phone
	}
	if req.Email != "" {
		company.Email = req.Email
	}
	if req.NPWP != "" {
		company.NPWP = req.NPWP
	}
	if req.NPP != "" {
		company.NPP = req.NPP
	}
	if req.Logo != "" {
		company.Logo = req.Logo
	}
	if req.IsActive != nil {
		company.IsActive = *req.IsActive
	}

	if err := s.companyRepo.Update(company); err != nil {
		return nil, errors.New("failed to update company")
	}

	response := dto.ToCompanyResponse(company)
	return &response, nil
}

func (s *companyService) Delete(id string) error {
	_, err := s.companyRepo.FindByID(id)
	if err != nil {
		return errors.New("company not found")
	}
	return s.companyRepo.Delete(id)
}

func (s *companyService) DeleteMultiple(ids []string) error {
	if len(ids) == 0 {
		return errors.New("no IDs provided")
	}
	return s.companyRepo.DeleteMultiple(ids)
}
